core/src/main/java/io/grpc/util/RoundRobinLoadBalancerFactory.java

```diff
@@ -186,23 +194,40 @@ public void handleNameResolutionError(Status error) {
 
     @Override
     public void handleSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateInfo) {
-      if (stateInfo.getState() == SHUTDOWN && stickinessState != null) {
+      Ref<ConnectivityStateInfo> stateInfoRef = getSubchannelStateInfoRef(subchannel);
+      ConnectivityState stateBefore = stateInfoRef.value.getState();
+      if (stateBefore == SHUTDOWN) {
+        return;
+      }
+      ConnectivityState newState = stateInfo.getState();
+      if (newState == SHUTDOWN && stickinessState != null) {
         stickinessState.remove(subchannel);
       }
       if (subchannels.get(subchannel.getAddresses()) != subchannel) {
         return;
       }
-      if (stateInfo.getState() == IDLE) {
+      if (newState == IDLE) {
         subchannel.requestConnection();
       }
-      getSubchannelStateInfoRef(subchannel).value = stateInfo;
-      updateBalancingState(getAggregatedState(), getAggregatedError());
+      stateInfoRef.value = stateInfo;
+      if (!ready || (newState == READY ^ stateBefore == READY)) {
+        updateBalancingState(null, null);
```

Comment:
nit: we generally avoid using literals that are not obvious.  IN this case, consider naming the literal nulls like:

`updateBalancingState(/*connectivityState= */ null, /* status= */ null);`